src/spaniel-instance.ts

```diff
@@ -0,0 +1,199 @@
+/*
+Copyright 2017 LinkedIn Corp. Licensed under the Apache License,
+Version 2.0 (the "License"); you may not use this file except in
+compliance with the License. You may obtain a copy of the License
+at http://www.apache.org/licenses/LICENSE-2.0
+ 
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*/
+
+import {
+  OptimizedEngineInterface,
+  OptimizedEngine,
+  Engine
+} from './metal/engine';
+
+import {
+  SpanielObserver
+} from './spaniel-observer';
+
+import {
+  SpanielIntersectionObserver,
+  generateEntry
+} from './intersection-observer';
+
+import {
+  entrySatisfiesRatio
+} from './utils';
+
+import {
+  SpanielTrackedElement,
+  SpanielObserverEntry,
+  DOMString,
+  DOMMargin,
+  IntersectionObserverClass,
+  SpanielObserverInit,
+  IntersectionObserverInit
+} from './interfaces';
+
+import {
+  EventStore,
+  RAFEventRecord,
+  GenericEventRecord
+} from './metal/events';
+
+import {
+  Scheduler,
+  Frame,
+  generateToken,
+  ElementScheduler
+} from './metal/index';
+
+import {
+  FrameInterface
+} from './metal/interfaces';
+import { Watcher, WatcherConfig } from './watcher';
+
+import w from './metal/window-proxy';
+
+let IntersectionObserver: IntersectionObserverClass = !!w.IntersectionObserver ? w.IntersectionObserver : undefined;
+
+export interface SpanielIntersectionObserverInterface {
+  new(callback: Function, options: IntersectionObserverInit): SpanielIntersectionObserver;
+}
+
+function generateIntersectionObserverClass(parent: SpanielInstance): SpanielIntersectionObserverInterface {
+  return class SpanielIntersectionObserverClass extends SpanielIntersectionObserver {
+    constructor(callback: Function, options: IntersectionObserverInit = {}) {
+      super(callback, options, parent);
+    }
+  };
+}
+
+export interface SpanielObserverInterface {
+  new (callback: (entries: SpanielObserverEntry[]) => void, options: SpanielObserverInit): SpanielObserver;
+}
+
+function generateSpanielObserver(parent: SpanielInstance): SpanielObserverInterface {
+  return class SpanielObserverClass extends SpanielObserver {
+    constructor(callback: (entries: SpanielObserverEntry[]) => void, options: SpanielObserverInit = {}) {
+      super(callback, options, parent);
+    }
+  };
+}
+
+export interface WatcherInterface {
+  new(config: WatcherConfig): Watcher;
+}
+
+function generateWatcher(parent: SpanielInstance): WatcherInterface {
+  return class WatcherClass extends Watcher {
+    constructor(config: WatcherConfig = {}) {
+      super(config, parent);
+    }
+  };
+}
+
+export class SpanielInstance {
```

Comment:
yes, I cry each time i re-encounter ApplicationInstance instances.... We shouldn't repeat this mistake here..

As spaniel is the namespace. 

* `Spaniel.Session` <-- nice and general, but not too word soup.
* `Spaniel.Factory` <-- yuck, but better then having instance instances 📦